pkg/collector/corechecks/system/wincrashdetect/probe/crashparse.go

```diff
@@ -0,0 +1,197 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+//go:build windows
+// +build windows
+
+package probe
+
+/*
+#cgo LDFLAGS: -l dbgeng -static
+#include "crashdump.h"
+*/
+import "C"
+import (
+	"fmt"
+	"strings"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// allow us to change for testing
+var readfn = doReadCrashDump
+
+type logCallbackContext struct {
+	loglines       []string
+	hasSeenRetAddr bool
+	unfinished     string
+}
+
+// maximum number of stack trace lines we'll look through, looking for non-"NT!" lines
+const maxLinesToScan = int(200)
+
+const (
+	bugcheckCodePrefix     = "Bugcheck code"
+	debugSessionTimePrefix = "Debug session time"
+	retAddrPrefix          = "RetAddr"
+	unableToLoadPrefix     = "Unable to"
+	ntBangPrefix           = "nt!"
+)
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+/*
+ * extra layer of indirection so that we can call the go parsing code
+ * (logLineCallbackGo) straight from the test function so that we can
+ * test out the parser with known input rather than actually calling
+ * the debugger
+ */
+//export logLineCallback
+func logLineCallback(voidctx C.PVOID, str C.PCSTR) {
+	ctx := (*logCallbackContext)(unsafe.Pointer(uintptr(voidctx)))
+	line := C.GoString(str)
+	logLineCallbackGo(ctx, line)
+}
+
+func logLineCallbackGo(ctx *logCallbackContext, line string) {
+	if !strings.Contains(line, "\n") {
+		ctx.unfinished = ctx.unfinished + line
+		return
+	}
+	if len(ctx.unfinished) != 0 {
+		line = ctx.unfinished + line
+		ctx.unfinished = ""
+	}
+	lines := strings.Split(line, "\n")
+
+	// if the last line is _not_ empty, that means it did not end with a `\n`.  So save that
+	// away for the next round
+	numlines := len(lines)
+	if len(lines[numlines-1]) != 0 {
+		ctx.unfinished = lines[numlines-1]
+		lines[numlines-1] = ""
+	}
+	start := int(0)
+	if !ctx.hasSeenRetAddr {
+		for idx, l := range lines {
+			if strings.HasPrefix(l, bugcheckCodePrefix) {
+				ctx.loglines = append(ctx.loglines, l)
+			}
+			if strings.HasPrefix(l, debugSessionTimePrefix) {
+				ctx.loglines = append(ctx.loglines, l)
+			}
+			if strings.HasPrefix(l, retAddrPrefix) {
+				ctx.hasSeenRetAddr = true
+				start = idx
+			}
+		}
+		if !ctx.hasSeenRetAddr {
+			return
+		}
+
+	}
+	ctx.loglines = append(ctx.loglines, lines[start:]...)
+}
+
+// this extra layer of indirection so that we can swap out test code which skips the actual debugger.
+func doReadCrashDump(filename string, ctx *logCallbackContext, exterr *uint32) error {
+	err := C.readCrashDump(C.CString(filename), unsafe.Pointer(ctx), (*C.long)(unsafe.Pointer(exterr)))
```

Comment:
You need to free the result of `C.CString` at some point.

Reply:
Can you test your program while using `GODEBUG=cgocheck=2` as an env var? I'm unsure if the pointer passing from Go->C->Go is done correctly. You may want to use a `cgo.Handle`.

Reply:
I ran as suggested, and the result was no error.